### PR TITLE
RavenDB-13110 - Sharding - external and pull replication integration

### PR DESCRIPTION
**Pull request by shiranshalom**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-13110

### Additional description

Continuation of https://github.com/ravendb/ravendb/pull/14372 (`ReplicationLoader `abstraction)

### Type of change

- New feature

### How risky is the change?

- Not relevant

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs [hunk 1]
```diff
@@ -326,19 +318,13 @@ internal class MergedDocumentReplicationCommand : TransactionOperationsMerger.Me
         {
             private readonly long _lastEtag;
             private readonly DataForReplicationCommand _replicationInfo;
-            private readonly ReplicationLatestEtagRequest.ReplicationType _replicationType;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs:329]: check the warning in `public LiveReplicationPerformanceCollector.ReplicationPerformanceType GetReplicationPerformanceType()`

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs (hunk 1 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs:329]: done

File: src/Raven.Server/Documents/Replication/AbstractReplicationLoader.cs [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+
+        protected AbstractReplicationLoader(ServerStore serverStore, string databaseName, RavenConfiguration configuration)
+        {
+            _databaseName = databaseName;
+            _server = serverStore;
+            _logger = LoggingSource.Instance.GetLogger(GetType().FullName, databaseName);
+
+            var config = configuration.Replication;
+            var reconnectTime = config.RetryReplicateAfter.AsTimeSpan;
+            _reconnectAttemptTimer = new Timer(state => ForceTryReconnectAll(),
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/AbstractReplicationLoader.cs]: should this be in the non sharded only?

File: src/Raven.Server/Documents/Replication/AbstractReplicationLoader.cs [hunk 3]
```diff
@@ -0,0 +1,541 @@
...
+
+        protected bool IsMyTask(Dictionary<string, RavenConnectionString> connectionStrings, DatabaseTopology topology, ExternalReplicationBase task)
+        {
+            if (ValidateConnectionString(connectionStrings, task, out _) == false)
+                return false;
+
+            var taskStatus = GetExternalReplicationState(_server, _databaseName, task.TaskId);
+            var whoseTaskIsIt = _server.WhoseTaskIsIt(topology, task, taskStatus);
+            return whoseTaskIsIt == _server.NodeTag;
+        }
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/AbstractReplicationLoader.cs]: wonder if this method shouldn't be abstract.. how would that work in the sharded version?

File: src/Raven.Server/Documents/Replication/AbstractReplicationLoader.cs [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+        {
+            if (_server.IsPassive())
+            {
+                throw new InvalidOperationException(
+                    $"Cannot accept the incoming replication connection from {connectionInfo.SourceUrl}, because this node is in passive state.");
+            }
+        }
+
+        protected virtual DynamicJsonValue GetInitialRequestMessage(ReplicationLatestEtagRequest replicationLatestEtagRequest,
+            ReplicationLoader.PullReplicationParams replParams = null)
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/AbstractReplicationLoader.cs:304]: can we avoid the `ReplicationLoader.PullReplicationParams replParams = null` here ?

File: src/Raven.Server/Documents/Replication/AbstractReplicationLoader.cs [hunk 5]
```diff
@@ -0,0 +1,541 @@
...
+        }
+
+        protected DatabaseTopology GetTopologyForShard(int shard)
+        {
+            using (_server.ContextPool.AllocateOperationContext(out TransactionOperationContext ctx))
+            using (ctx.OpenReadTransaction())
+            {
+                return _server.Cluster.ReadDatabaseTopologyForShard(ctx, ShardHelper.ToDatabaseName(_databaseName), shard);
+            }
+        }
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/AbstractReplicationLoader.cs:332]: this should be in the sharded version no?

File: src/Raven.Server/Documents/Replication/AbstractReplicationLoader.cs [hunk 6]
```diff
@@ -0,0 +1,541 @@
...
+                _outgoing.Clear();
+                _incoming.Clear();
+            }
+            finally
+            {
+                _locker.ExitWriteLock();
+            }
+        }
+
+        public class ConnectionShutdownInfo
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/AbstractReplicationLoader.cs]: move to new file

File: src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs [hunk 7]
```diff
@@ -360,6 +364,13 @@ public IncomingReplicationStatsAggregator GetLatestReplicationPerformance()
             return _lastStats;
         }
 
+        public LiveReplicationPerformanceCollector.ReplicationPerformanceType GetReplicationPerformanceType()
+        {
+            return ReplicationType == ReplicationLatestEtagRequest.ReplicationType.Internal
+                ? LiveReplicationPerformanceCollector.ReplicationPerformanceType.IncomingInternal
+                : LiveReplicationPerformanceCollector.ReplicationPerformanceType.IncomingExternal;
+        }
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs]: Can this be implemented in the appropriate IncomingReplicationHandlers ?

File: src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs [hunk 8]
```diff
@@ -85,6 +88,7 @@ public abstract class AbstractIncomingReplicationHandler<TContextPool, TOperatio
             ConnectionInfo = IncomingConnectionInfo.FromGetLatestEtag(replicatedLastEtag);
             SupportedFeatures = TcpConnectionHeaderMessage.GetSupportedFeaturesFor(tcpConnectionOptions.Operation, tcpConnectionOptions.ProtocolVersion);
             ConnectionInfo.RemoteIp = ((IPEndPoint)_tcpClient.Client.RemoteEndPoint)?.Address.ToString();
+            ReplicationType = replicationType;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs]: do we need that only for `GetReplicationPerformanceType` ?

File: src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs [hunk 9]
```diff
@@ -685,7 +663,7 @@ protected override long ExecuteCmd(DocumentsOperationContext context)
             protected virtual void SaveSourceEtag(DocumentsOperationContext context)
             {
                 context.LastReplicationEtagFrom ??= new Dictionary<string, long>();
-                if (_replicationType != ReplicationLatestEtagRequest.ReplicationType.Sharded)
+                if (_replicationInfo.SourceDatabaseId != null)
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs:681]: hmm we use the `IncomingReplicationHandler` for both sharded and non-sharded replication?

File: src/Raven.Server/Documents/Replication/AbstractReplicationLoader.cs (hunk 5 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/AbstractReplicationLoader.cs:332]: No. It was in `ReplicationLoader` (we are using it in `HandleMigrationReplication` method)

File: src/Raven.Server/Documents/Replication/AbstractReplicationLoader.cs (hunk 6 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/AbstractReplicationLoader.cs]: done

File: src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs (hunk 9 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs:681]: Yep

File: src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs (hunk 7 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs]: Umm.. I don't see where we can handle it (we don't have `IncomingExternalReplicationHandler` for example)

File: src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs (hunk 8 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs]: Yes.. any idea how to avoid it? :-)

File: src/Raven.Server/Documents/Replication/AbstractReplicationLoader.cs (hunk 2 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/AbstractReplicationLoader.cs]: done

File: src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs (hunk 8 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs]: done

File: src/Raven.Server/Documents/Replication/AbstractReplicationLoader.cs (hunk 3 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Replication/AbstractReplicationLoader.cs]: Done - moved back to `ReplicationLoader`

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs [hunk 10]
```diff
@@ -122,6 +133,23 @@ private void PrepareReplicationDataForShards(TransactionOperationContext context
             }
         }
 
+        private byte[] GetBufferFromAttachmentStream(Stream stream)
+        {
+            var length = (int)stream.Length;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs:138]: would that break if the attachment is >2GB?

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs [hunk 11]
```diff
@@ -122,6 +133,23 @@ private void PrepareReplicationDataForShards(TransactionOperationContext context
...
+            var buffer = new byte[length];
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs:139]: this is pure pain :D

File: src/Raven.Server/Documents/Handlers/Processors/Replication/PullReplicationHandlerProcessorForDefineHub.cs [hunk 12]
```diff
@@ -8,5 +9,12 @@ internal class PullReplicationHandlerProcessorForDefineHub : AbstractPullReplica
         public PullReplicationHandlerProcessorForDefineHub([NotNull] DatabaseRequestHandler requestHandler) : base(requestHandler)
         {
         }
+
+        protected override ValueTask AssertCanExecuteAsync()
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Handlers/Processors/Replication/PullReplicationHandlerProcessorForDefineHub.cs:13]: why was this moved form the abstract?

File: src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs [hunk 13]
```diff
@@ -484,6 +487,8 @@ protected void ReceiveSingleDocumentsBatch(TOperationContext context, int replic
                             while (task.Wait(Math.Min(3000, (int)(configuration.Replication.ActiveConnectionTimeout.AsTimeSpan.TotalMilliseconds * 2 / 3))) ==
                                    false)
                             {
+                                HandleMissingAttachmentsIfNeeded(ref task);
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/Incoming/AbstractIncomingReplicationHandler.cs:490]: what is that??

File: src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs (hunk 9 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/Replication/Incoming/IncomingReplicationHandler.cs:681]: add a dev helper to fix this as part of the sharded ext. failover

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs [hunk 14]
```diff
@@ -667,189 +598,38 @@ public enum ConnectionType
...
             {
-                DatabaseTopology topology;
-                Dictionary<string, RavenConnectionString> ravenConnectionStrings;
+                string changeVector = null;
+                long lastEtagFromSrc = 0;
 
-                using (_server.ContextPool.AllocateOperationContext(out TransactionOperationContext ctx))
-                using (ctx.OpenReadTransaction())
+                if (getLatestEtagMessage.ReplicationsType != ReplicationLatestEtagRequest.ReplicationType.Migration &&
+                    ShardHelper.IsShardedName(Database.Name) == false)
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/ReplicationLoader.cs:614]: Please add a todo to check if we can simplify the code here and always send the change vector/etag?

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs [hunk 15]
```diff
@@ -667,189 +598,38 @@ public enum ConnectionType
...
-                    $"Initialized document replication connection from {connectionInfo.SourceDatabaseName} located at {connectionInfo.SourceUrl}");
-
-            return getLatestEtagMessage;
-        }
-
-        private long _reconnectInProgress;
+        internal static readonly TimeSpan MaxInactiveTime = TimeSpan.FromSeconds(60);
 
-        private void ForceTryReconnectAll()
+        protected override DynamicJsonValue GetInitialRequestMessage(ReplicationLatestEtagRequest getLatestEtagMessage, PullReplicationParams replParams = null)
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/ReplicationLoader.cs:603]: the name should be `GetResponseToReplicationLatestEtagRequest` or something like that? you are not getting here the initial request message, right?

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs [hunk 16]
```diff
@@ -1185,7 +1054,7 @@ private void DisposeConnections(List<IDisposable> instancesToDispose)
 
                     // ReSharper disable once PossibleUnintendedReferenceComparison
                     var handler = outgoingHandlers.Find(o => o.Destination == actual); // we explicitly compare references.
-                    if (handler == null)
+                    if (handler == null || handler is OutgoingReplicationHandlerBase outgoingHandler == false)
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/ReplicationLoader.cs:1057]: `OutgoingReplicationHandlerBase` vs `AbstractOutgoingReplicationHandler` is _very_ confusing.... 
let's rename the `OutgoingReplicationHandlerBase`  to `DocumentDatabaseOutgoingReplicationHandler`

Comment (karmeli87) [reply src/Raven.Server/Documents/Replication/ReplicationLoader.cs:1057]: and why is the condition above?

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedDatabaseRequestHandler.cs [hunk 17]
```diff
@@ -65,12 +65,16 @@ public override void Init(RequestHandlerContext context)
 
             var request = HttpContext.Request;
             var url = request.Path.Value;
-            var relativeIndex = url.IndexOf('/', 11); //start after "/databases/" and skip the database name
 
-            BaseShardUrl = url.Substring(relativeIndex);
-            RelativeShardUrl = BaseShardUrl + request.QueryString;
-            Method = new HttpMethod(request.Method);
+            if (string.IsNullOrEmpty(url) == false)
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedDatabaseRequestHandler.cs:69]: when can this happened?

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs [hunk 18]
```diff
@@ -144,11 +142,12 @@ private bool SendDocumentsBatch(TransactionOperationContext context, List<Replic
                 _stream.Flush();
                 sw.Stop();
 
-                var (type, _) = HandleServerResponse();
+                var (type, reply) = HandleServerResponse(getFullResponse: true);
 
                 if (type == ReplicationMessageReply.ReplyType.MissingAttachments)
                 {
                     MissingAttachmentsInLastBatch = true;
+                    _replicationQueue.MissingAttachmentMessage = reply?.Exception;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs:150]: `MissingAttachmentMessage`->`MissingAttachmentException`

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs [hunk 19]
```diff
@@ -51,7 +49,7 @@ protected override void Replicate()
         {
             while (_cts.IsCancellationRequested == false)
             {
-                while (_replicationQueue.Items[_shardNumber].TryTake(out var items))
+                while (_replicationQueue.Items[_shardNumber].TryTake(out var items, 0, _cts.Token))
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs:52]: if there is nothing if the queue you will have here a tight loop

Comment (karmeli87) [reply src/Raven.Server/Documents/Sharding/Handlers/ShardedOutgoingReplicationHandler.cs:52]: you need to use here the `Take` overload

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 20]
```diff
@@ -3505,10 +3505,12 @@ public DatabaseTopology ReadDatabaseTopology<TTransaction>(TransactionOperationC
         {
             using (var databaseRecord = ReadRawDatabaseRecord(context, name))
             {
+                DatabaseTopology topology;
                 if (databaseRecord.IsSharded)
-                    throw new InvalidOperationException($"The database record '{name}' is sharded and doesn't contain topology directly.");
+                    topology = databaseRecord.Sharding?.Orchestrator.Topology;
```
Comment (karmeli87) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs:3510]: why this change?

**Comment by shiranshalom** [comment]
Closing this. New PR: https://github.com/ravendb/ravendb/pull/14652
